### DISCUSSION TITLE
New Feature - Allowing Editing of Questions (New + reused content) by reviewers

### DISCUSSION THREAD

**Discussion by Prerna789** [Sunbird-coKreat/Community#8] [Feature Requests]
Hi All,

I am sharing a brief note on the new feature that we have planned as part of Prashnavali solution on VDN.

**Functionality** - Allowing reviewers to make edits to the questions (text, options, metadata fields)

**Challenges in the current solution flow -** 

- User persona of question contributions - Subject teachers with low/medium comfort with tech
- Purposes for question creation under each project - Assessment creation, worksheet creation etc
- Time duration for finalisation of questions (with review) in a project - 2/3 weeks

Currently, a teacher who is assigned a project is provided with a target of making 50/60 question contributions within 1/2 weeks. 
Minor mistakes in the question creation process such as spelling errors, grammatically incorrect sentences, incorrect order of sentences etc. are commonly made by users.

The questions are then sent for review.
For reviewers to have to send back the question for correction for minor changes which might not be related to pedagogy or quality, there is a high turnaround time.

This does not work well for curation of content for shorter projects and time can be optimized if the reviewer has access to make minor changes to the content or structure of the question.

**Functional solution -** 
1. Allowing contribution and sourcing reviewer to make edits to a question during review (for new content that is created)
2. Allowing reviewers to make a copy of question and make edits to the copy (if reused content is edited)

Detailed PRD along with the Tech design is updated in the doc - https://project-sunbird.atlassian.net/wiki/spaces/PRD/pages/2884567041/Correction+of+a+content+by+sourcing+reviewer

Would request you to go through it and share your comments 
@Krishnaj20 @pallakartheekreddy [@vrayulu](https://github.com/vrayulu)
[@maheshkumargangula](https://github.com/maheshkumargangula)

**Comment by Prerna789** [comment]
[@vrayulu](https://github.com/vrayulu)
[@maheshkumargangula](https://github.com/maheshkumargangula)
Request you to go through the design and provide feedback.
Thanks
cc - @Krishnaj20

**Reply by Krishnaj20** [reply]
@maheshkumargangula - can you share your feedback to the design proposed for this change please?

**Reply by maheshkumargangula** [reply]
@Prerna789 - Below is my feedback for the proposed design.

1. Without the creator's knowledge the reviewer should not be allowed to make modifications. May be, while sending the Question for review, the creator can add one of the reviewer as collaborator explicitly to make modifications by only that reviewer.
2. Storing the access control information along with metadata is not good. It will be easy to make changes to it and this information flow everwhere (even in consumption scenarios also). We should separate the access control data from metadata while saving. 
3. We should generate the relevant telemetry when we are making changes to access or manage the object or adding the collaborators to it. Please define the required telemetry changes also in the tech design.

Cc: @Krishnaj20 @vrayulu

**Reply by pramodkvarma** [reply]
I agree with most of what @maheshkumargangula suggested except for one thing. 

On point 1, I suggest we allow a ACL config that says "allow reviewers to make changes if necessary during the review" and default it to "yes". If creators do not want it let them uncheck it. 

Adding reviewers as collaborators may be tricky since many cases creators (especially crowd sourcing type situation) may not know exactly who the reviewer will be. Hence I am suggesting the above.

Agree fully with points 2 and 3 made by Mahesh.

**Reply by kameshbhr** [reply]
I agree with Pramod. Creators adding the reviewers will be tricky. 

Here is one thought that got triggered based on Pramod's comment. PRD talks about enabling this per project basis and only for private projects (contributions coming from same org). We should probably make it more generic and simpler. Why can't we just have this enabled for any contribution? If the contributor/creator belongs to the same org, the option "Allow reviewers to make changes..." will be set to "Yes" by default. If the contributor doesn't belong to the same org, the option will be set to "No" by default. 
At a system/tenant level have a config to turn on this feature or not. If the feature is not turned on, by default reviewers will not be able to edit.

We should also enable notification to be sent to the creator when reviewer modifies the content.

**Reply by pramodkvarma** [reply]
Yes @kameshbhr, agree fully.

**Reply by Krishnaj20** [reply]
thanks @pramodkvarma @maheshkumargangula @kameshbhr for your feedback. Summarizing the recommended approach:
a) 'Allow reviewers to edit content' - can be enabled/disabled at a system level. This configuration can be overriden at a tenant level.
b) When this configuration is turned ON, then the contributors are informed (at the time of contribution) that reviewers can make changes to their contribution.
c) If this configuration is turned ON, then the reviewers will be able to make changes to the contribution and publish the contribution.

@Prerna789 - Can you please relook at the design to accommodate these comments? I understand that you have already started with the implementation from this week but since the build for 5.0 has just started from this Monday, I'm hoping there is not much of re-work needed.

**Reply by vrayulu** [reply]
@Krishnaj20 @Prerna789 - what about "creator can disallow reviewers to edit for a specific asset"? i guess we should allow that also.

**Reply by kameshbhr** [reply]
@Krishnaj20 Few corrections based on my understanding:
a) 'Allow reviewers to edit content' - can be enabled/disabled at a system level. This configuration can be overriden at a tenant level.
b) When this configuration is turned ON, then the contributors are shown a check box with a message "Allow reviewers to make changes to this content" (at the time of contribution). This is by default checked, if the contributor belongs to the same org to which he/she is contributing to. It is by default unchecked, if the contributor doesn't belong to the same org to which he/she is contributing. User's choice is stored in the system.
c) The reviewers will be able to make changes to the contribution for which the user's choice was to allow reviewers to make changes.

This also answers @vrayulu's question.

**Reply by Krishnaj20** [reply]
thanks @kameshbhr & @vrayulu . I agree with your comments.

@Prerna789 - both the PRD & tech design will need changes to accommodate this feedback. Please do share the updated artefacts in this thread. Feel free to reach out to me & @pallakartheekreddy - we can assist you to create these artefacts.

**Reply by manojSRawat** [reply]
We've updated the PRD based on the above feedback.

I've following concern for the above suggestions.
1) consent default check (@kameshbhr): 
Consent is something we want users to be agree on, making it default select means we are forcing the user.
We can give an accept-all checkbox at the top as shown in the mockups.

2) Storing access control in metadata (@maheshkumargangula): 
     Currently, we can add question in a question-set by 2 methods
     1. By creating a new question
     2. We can add existing questions from the library
     
      In 2nd method, if we add an existing question then we can only access the question object and if we don't store access control to question then we never know the editability of the question.
      
cc: @Prerna789

**Reply by Prerna789** [reply]
Hi @surendrasinghs @pallakartheekreddy @maheshkumargangula @kameshbhr 

Manoj had updated the PRD and added his comments on your feedback.
We are going ahead with implementation of the feature basis the updated PRD Doc, considering it as sign off. 

Thanks!
cc - @manojSRawat
